cmd/thanos/query.go

```diff
@@ -127,7 +142,52 @@ func storeClientGRPCOpts(reg *prometheus.Registry, tracer opentracing.Tracer) []
 		reg.MustRegister(grpcMets)
 	}
 
-	return dialOpts
+	if !secure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+		return dialOpts, nil
+	}
+
+	level.Info(logger).Log("msg", "Enabling client to server TLS")
+
+	var certPool *x509.CertPool
+
+	if caCert != "" {
+		caPEM, err := ioutil.ReadFile(caCert)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading client CA")
+		}
+
+		certPool = x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(caPEM) {
+			return nil, errors.Wrap(err, "building client CA")
+		}
+		level.Info(logger).Log("msg", "TLS Client using provided certificate pool")
+	} else {
+		var err error
+		certPool, err = x509.SystemCertPool()
```

Comment:
nice one. Wonder if we can do same for server. Grab those from system and add server flag for `grpc-server-client-cert-required`.. not sure (:

Reply:
you can see kind of not perfect example here: https://github.com/improbable-eng/kedge/blob/master/cmd/kedge/tls.go#L26

Reply:
I'm not sure if that's hugely meaningful though, it feels odd to say that any publicly signed TLS cert could be used as a client. It feels like this should always be at least restricted to an organisation's issuing CA, if even that CA happens to be producing certs that the whole world could trust. And, if that really is intended, they can just manually point at the local machines ca-certificates.pem (hopefully the act of adding that flag will cause them to think about what they are doing, and decide against it).